### ISSUE TITLE
Feature Request: Simulating Larger Batch Sizes

### ISSUE THREAD

**Issue by rsomani95** [mlfoundations/open_clip#229]
I was wondering if it's possible to simulate having larger batches by storing image and text features until we wait to do an update. 

Specifically, if the GPU allows us to have a `batch_size` of `8` but we want a desired batch size of `32`, we compute the loss only at the end of every 4th batch.

As contrastive learning benefits greatly from larger batch sizes, having this feature would be invaluable for fine-tuning cases that could be done on a single GPU that only allows for a small batch size

**Comment by rom1504** [comment]
you may want to try this PR https://github.com/mlfoundations/open_clip/pull/108

**Comment by usuyama** [comment]
Are you already using the gradient checkpointing and local loss features? Those can reduce the memory usage quite a lot `--grad-checkpointing --local-loss  --gather-with-grad`

**Comment by rsomani95** [comment]
@usuyama I was using `--local-loss` and `--gather-with-grad`, but wasn't aware of `grad-checkpointing`. That's pretty handy, I was able to increase batch size from `24` to `224` per GPU! Thanks

@rom1504 That looks interesting. I'll check it out. Do you have any clue if there's plans to merge that PR, or would that functionality be enabled later via a re-write of #74 (cc @rwightman)

**Comment by rwightman** [comment]
@rsomani95 I was working w/ the grad caching a bit, but it is a bit messy and I was never satifisfied it was working 100% as it should. It also puts a heavy load on the CPU memory per train process. In the end it got pushed very low down priority list, this training technique works best with larger model and data scale, if you are in a position where grad caching did work it'd be prohibitivly long in wall clock time to rely on it, you really just need a lot of GPUs :/

**Comment by mitchellnw** [comment]
You may be interested in https://github.com/mlfoundations/open_clip/pull/267
